modules/join/src/main/java/org/apache/ignite/internal/join/ClusterManagementGroupManager.java

```diff
@@ -0,0 +1,198 @@
+
+package org.apache.ignite.internal.join;
+
+import static org.apache.ignite.internal.join.Utils.resolveNodes;
+
+import io.netty.handler.codec.http.HttpHeaderValues;
+import io.netty.handler.codec.http.HttpMethod;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.ignite.internal.join.messages.CancelInitMessage;
+import org.apache.ignite.internal.join.messages.CmgInitMessage;
+import org.apache.ignite.internal.join.messages.InitMessageGroup;
+import org.apache.ignite.internal.join.messages.InitMessagesFactory;
+import org.apache.ignite.internal.join.rest.InitCommandHandler;
+import org.apache.ignite.internal.manager.IgniteComponent;
+import org.apache.ignite.internal.raft.Loza;
+import org.apache.ignite.internal.rest.RestModule;
+import org.apache.ignite.internal.rest.routes.Route;
+import org.apache.ignite.internal.util.IgniteSpinBusyLock;
+import org.apache.ignite.lang.IgniteInternalException;
+import org.apache.ignite.lang.IgniteLogger;
+import org.apache.ignite.lang.NodeStoppingException;
+import org.apache.ignite.network.ClusterNode;
+import org.apache.ignite.network.ClusterService;
+import org.apache.ignite.network.MessagingService;
+import org.apache.ignite.network.NetworkMessage;
+import org.apache.ignite.raft.client.Peer;
+
+/**
+ * Ignite component responsible for cluster initialization and managing the Cluster Management Raft Group.
+ */
+public class ClusterManagementGroupManager implements IgniteComponent {
+    private static final IgniteLogger log = IgniteLogger.forClass(ClusterManagementGroupManager.class);
+
+    /** CMG Raft group name. */
+    private static final String CMG_RAFT_GROUP_NAME = "cmg_raft_group";
+
+    /** Init REST endpoint path. */
+    private static final String REST_ENDPOINT = "/management/v1/cluster/init";
+
+    /** Busy lock to stop synchronously. */
+    private final IgniteSpinBusyLock busyLock = new IgniteSpinBusyLock();
+
+    /** Prevents double stopping the component. */
+    private final AtomicBoolean stopGuard = new AtomicBoolean();
+
+    private final ClusterService clusterService;
+
+    private final Loza raftManager;
+
+    private final RestModule restModule;
+
+    /** Handles cluster initialization flow. */
+    private final ClusterInitializer clusterInitializer;
+
+    /** Constructor. */
+    public ClusterManagementGroupManager(ClusterService clusterService, Loza raftManager, RestModule restModule) {
+        this.clusterService = clusterService;
+        this.raftManager = raftManager;
+        this.restModule = restModule;
+        this.clusterInitializer = new ClusterInitializer(clusterService);
+
+        MessagingService messagingService = clusterService.messagingService();
+
+        var msgFactory = new InitMessagesFactory();
+
+        messagingService.addMessageHandler(InitMessageGroup.class, (msg, addr, correlationId) -> {
```

Comment:
> The rule of thumb is to prefer start() to constructor for registering message handlers

I would argue with that, because I actually think that it is better to register handlers in constructors instead. Here are my reasons:

If message handlers are registered in "start", then the network layer is already running, and we can miss some messages. Yes, the node might not have completely started yet, but at least we can either block or return a meaningful error. Otherwise these messages will simply be dropped.

Reply:
There are two main points to keep constructors simple:
- Exception handling. If something went wrong during start, there's always a component provided by constructor as an entry point for cleanup or recovery. In other words there's always an object on which to call close
```
IgniteComponent c = new Component() // No tough logic goes here;
try {
c.start();
} catch(Exception e) {
c.close(); // Cleaning all resources that might be instantiated; 
}
```
- In general there are no guaranties that un-started components will be able to register listeners etc. I'd rather prefer simple bifurcation of full-fledged started component, and only-instantiated which is only suitable for calling the start method. That'll also ease migration to DI a bit.

> If message handlers are registered in "start", then the network layer is already running, and we can miss some messages. 
Seems that currently it's not possible. If it is, let's introduce onBeforeStart() or similar.

Reply:
> Exception handling

I don't think this is applicable to registering listeners

> In general there are no guaranties that un-started components will be able to register listeners etc. I'd rather prefer simple bifurcation of full-fledged started component, and only-instantiated which is only suitable for calling the start method. That'll also ease migration to DI a bit.

Why and how?

> Seems that currently it's not possible.

I think that it is still possible, since `ClusterService` is started before all other components

Reply:
> I don't think this is applicable to registering listeners
Why not? It's still possible to get an exception on listeners registering. My main point is to keep constructors as simple as possible moving all logic to start() or beforeStart() if it's required.
>Why and how?
Well, because it's not declared anywhere. I actually believe that concept of fully-fledged component after corresponding start is still simplier to use and understand.
>I think that it is still possible, since ClusterService is started before all other components
I mean, that currently there are no such cases.